pkg/cgroups/cgroupid.go

```diff
@@ -0,0 +1,78 @@
+package cgroups
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+// CgroupID implements mapping kernel cgroup IDs to cgroupfs paths with transparent caching.
+type CgroupID struct {
+	root  string
+	cache map[uint64]string
+	sync.Mutex
+}
+
+// NewCgroupID creates a new CgroupID map/cache.
+func NewCgroupID(root string) *CgroupID {
+	return &CgroupID{
+		root:  root,
+		cache: make(map[uint64]string),
+	}
+}
+
+func getID(path string) uint64 {
+	h, _, err := unix.NameToHandleAt(unix.AT_FDCWD, path, 0)
+	if err != nil {
+		return 0
+	}
+
+	return binary.LittleEndian.Uint64(h.Bytes())
+}
+
+// Find finds the path for the given cgroup id.
+func (cgid *CgroupID) Find(id uint64) (string, error) {
+	found := false
+	var p string
+
+	cgid.Lock()
+	defer cgid.Unlock()
+
+	if path, ok := cgid.cache[id]; ok {
+		return path, nil
+	}
+
+	err := fsi.Walk(cgid.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil
+			}
+			fmt.Printf("WalkFunc called with an error (path %q: %v\n)", path, err)
```

Comment:
Should we log instead?

Reply:
Good catch. I ended up removing this message. The cgroups package does not print anything anywhere else, and it acts as a lower level library for blockio. It didn't feel too good to establish new "[ cgroups ]" logger for this message alone.